File: scripts/ocr-worker.js
/**
 * OCR Worker — extract photo_timestamp from recent events
 * Runs every 5 min via GitHub Actions, processes events from the last 4 hours.
 */

import sharp from 'sharp';
import Tesseract from 'tesseract.js';

const SUPABASE_URL = process.env.SUPABASE_URL;
const SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

const HEADERS = {
  apikey: SERVICE_KEY,
  Authorization: `Bearer ${SERVICE_KEY}`,
  'Content-Type': 'application/json',
};

const RU_MONTHS = {
  'янв':1,'фев':2,'мар':3,'апр':4,'май':5,'мая':5,
  'июн':6,'июл':7,'авг':8,'сен':9,'окт':10,'ноя':11,'дек':12,
};
const EN_MONTHS = {
  'jan':1,'feb':2,'mar':3,'apr':4,'may':5,'jun':6,
  'jul':7,'aug':8,'sep':9,'oct':10,'nov':11,'dec':12,
};

function prepareText(raw) {
  return raw
    // OCR шум: ведущая "1" перед временем (" 123:55:30" → " 23:55:30")
    .replace(/\s1(\d{2}:\d{2}:\d{2})/g, ' $1')
    .replace(/(\d{2})1(\d{2}):(\d{2})/g, '$1:$2:$3')
    // Латинские OCR-замены для названий месяцев
    .replace(/(\d{1,2})\s*mas\s*(\d{4})/gi, '$1 мая $2')
    .replace(/(\d{1,2})\s*mai\s*(\d{4})/gi, '$1 мая $2')
    .replace(/(\d{1,2})\s*map\s*(\d{4})/gi, '$1 мар $2')
    .replace(/(\d{1,2})\s*waa\s*(\d{4})/gi, '$1 мая $2')   // "Waa" → "мая"
    .replace(/(\d{1,2})\s*was\s*(\d{4})/gi, '$1 мая $2')   // "Was" → "мая"
    .replace(/г[;,]/g, 'г.')
    .replace(/([а-яё]{3,4})\.\s*/gi, '$1 ')   // "июн. " → "июн "
    .replace(/(\d{4})\s*г\.\s*/g, '$1 ')       // "2026 г. " → "2026 "
    // Мусорная цифра/символ сразу после 4-значного года ("20261 "→"2026 ", "20260."→"2026.")
    .replace(/(\d{4})[01]([.\s,;])/g, '$1$2')
    // Точка сразу после года ("2026." → "2026 ")
    .replace(/(\d{4})\./g, '$1 ')
    // Изолированный "1" или "l" между годом и временем ("2026 1," → "2026 ")
    .replace(/(\d{4})\s+[1l][,\s]/gi, '$1 ');
}

function tryExtract(text) {
  const t = prepareText(text);

  // Формат: DD месяц YYYY HH:MM[:SS]
  const full = t.match(/(\d{1,2})\s*([а-яёa-z]{3,})\s*(\d{4})[^0-9]+(\d{1,2}):(\d{2})(?::(\d{2}))?/i);
  if (full) {
    const key  = full[2].toLowerCase().slice(0, 3);
    const mNum = RU_MONTHS[key] || EN_MONTHS[key];
    if (mNum) return { found: true, day: full[1], month: mNum, year: full[3], h: full[4], m: full[5], s: full[6] || '0' };
  }

  const time  = t.match(/(\d{1,2}):(\d{2}):(\d{2})/);  // HH:MM:SS
  const timeHM = t.match(/\b(\d{1,2}):(\d{2})\b/);      // HH:MM без секунд
  const dmy   = t.match(/(\d{1,2})[./](\d{1,2})[./](\d{2,4})/);
  const ymd   = t.match(/(\d{4})[-./](\d{2})[-./](\d{2})/);

  // Нормализуем 2-значный год (26 → 2026)
  function normalizeYear(y) { return y.length === 2 ? String(2000 + parseInt(y)) : y; }

  if (time && dmy) return { found: true, day: dmy[1], month: parseInt(dmy[2]), year: normalizeYear(dmy[3]), h: time[1], m: time[2], s: time[3] };
  if (time && ymd) return { found: true, day: ymd[3], month: parseInt(ymd[2]), year: ymd[1], h: time[1], m: time[2], s: time[3] };
  // HH:MM без секунд — только когда дата тоже найдена
  if (timeHM && dmy) return { found: true, day: dmy[1], month: parseInt(dmy[2]), year: normalizeYear(dmy[3]), h: timeHM[1], m: timeHM[2], s: '0' };
  if (timeHM && ymd) return { found: true, day: ymd[3], month: parseInt(ymd[2]), year: ymd[1], h: timeHM[1], m: timeHM[2], s: '0' };
  return { found: false };
}

function toIso({ day, month, year, h, m, s = '0' }) {
  const moscowMs = Date.UTC(parseInt(year), parseInt(month) - 1, parseInt(day),
                             parseInt(h), parseInt(m), parseInt(s));
  return new Date(moscowMs - 3 * 3600 * 1000).toISOString();
}

function isValidDate(isoStr) {
  const d = new Date(isoStr);
  if (isNaN(d.getTime())) return false;
  if (d > new Date()) return false;                    // будущее → ошибка OCR
  if (d < new Date('2026-01-01T00:00:00Z')) return false; // слишком старое → ошибка OCR
  return true;
}

async function ocrBuffer(buf, worker) {
  const r = await Promise.race([
    worker.recognize(buf),
    new Promise((_, rej) => setTimeout(() => rej(new Error('OCR timeout')), 15000)),
  ]).catch(() => null);
  return r?.data?.text || '';
}

async function extractOcrTimestamp(buffer, worker) {
  try {
    // Pass A: полное изображение
    const fullBuf = await sharp(buffer)
      .greyscale()
      .normalise()
      .resize({ width: 1600, withoutEnlargement: true, kernel: 'lanczos3' })
      .sharpen({ sigma: 1.5 })
      .toBuffer();

    await worker.setParameters({ tessedit_pageseg_mode: '6' }).catch(() => {});
    const resA = tryExtract(await ocrBuffer(fullBuf, worker));
    if (resA.found) return toIso(resA);

    // Pass B: угловые вырезки (правые 60% × верх/низ 30%)
    const meta  = await sharp(buffer).metadata();
    const cropL = Math.floor(meta.width * 0.40);
    const cropW = meta.width - cropL;
    const cornH = Math.floor(meta.height * 0.30);

    const makeCornerBuf = (top) =>
      sharp(buffer)
        .extract({ left: cropL, top, width: cropW, height: cornH })
        .greyscale()
        .normalise()
        .resize({ width: 900, kernel: 'lanczos3' })
        .sharpen({ sigma: 1.5 })
        .toBuffer();

    const resTop = tryExtract(await ocrBuffer(await makeCornerBuf(0), worker));
    if (resTop.found) return toIso(resTop);

    const resBot = tryExtract(await ocrBuffer(await makeCornerBuf(meta.height - cornH), worker));
    if (resBot.found) return toIso(resBot);

    // Pass C: нижняя полоса полной ширины — водяной знак в любом горизонтальном положении
    const stripTop = Math.floor(meta.height * 0.75);
    const stripH   = meta.height - stripTop;
    const resStrip = tryExtract(await ocrBuffer(
      await sharp(buffer)
        .extract({ left: 0, top: stripTop, width: meta.width, height: stripH })
        .greyscale().normalise()
        .resize({ width: 1200, kernel: 'lanczos3' })
        .sharpen({ sigma: 2 })
        .toBuffer(),
      worker
    ));
    if (resStrip.found) return toIso(resStrip);

    // Pass D: левый нижний угол, PSM 7 — Android-водяной знак слева
    await worker.setParameters({ tessedit_pageseg_mode: '7' }).catch(() => {});
    const resLeftBot = tryExtract(await ocrBuffer(
      await sharp(buffer)
        .extract({ left: 0, top: Math.floor(meta.height * 0.70),
                   width: Math.floor(meta.width * 0.60), height: Math.floor(meta.height * 0.30) })
        .greyscale().normalise()
        .resize({ width: 900, kernel: 'lanczos3' })
        .sharpen({ sigma: 2 })
        .toBuffer(),
      worker
    ));
    await worker.setParameters({ tessedit_pageseg_mode: '6' }).catch(() => {});
    if (resLeftBot.found) return toIso(resLeftBot);

    // Pass E: нижняя полоса с инверсией — тёмный текст на светлом фоне
    const resInv = tryExtract(await ocrBuffer(
      await sharp(buffer)
        .extract({ left: 0, top: stripTop, width: meta.width, height: stripH })
        .greyscale().normalise().negate()
        .resize({ width: 1200, kernel: 'lanczos3' })
        .sharpen({ sigma: 2 })
        .toBuffer(),
      worker
    ));
    if (resInv.found) return toIso(resInv);

    return null;
  } catch {
    return null;
  } finally {
    await worker.setParameters({ tessedit_pageseg_mode: '3' }).catch(() => {});
  }
}

async function downloadPhoto(photoUrl) {
  const url = `${SUPABASE_URL}/storage/v1/object/${photoUrl}`;
  const res = await fetch(url, { headers: HEADERS });
  if (!res.ok) throw new Error(`Storage download failed: ${res.status}`);
  return Buffer.from(await res.arrayBuffer());
}

async function updateEvent(eventId, photoTimestamp, currentFlags, eventStatus) {
  // Убираем no_photo_time — время успешно считано с фото
  const newFlags = (currentFlags || []).filter(f => f !== 'no_photo_time');
  const patch = {
    photo_timestamp: photoTimestamp,
    fraud_flags:     newFlags,
  };
  // Если событие уже обработано — сбрасываем в pending, AI worker пересчитает часы
  if (eventStatus === 'done') {
    patch.status = 'pending';
  }
  const res = await fetch(`${SUPABASE_URL}/rest/v1/events?id=eq.${eventId}`, {
    method: 'PATCH',
    headers: HEADERS,
    body: JSON.stringify(patch),
  });
  if (!res.ok) throw new Error(`Update failed: ${res.status}`);
}

async function main() {
  console.log('🔄 OCR Worker started');

  let worker = null;
  try {
    worker = await Tesseract.createWorker('rus+eng');

    const since = new Date(Date.now() - 72 * 60 * 60 * 1000).toISOString();
    const res = await fetch(
      `${SUPABASE_URL}/rest/v1/events?photo_url=not.is.null&photo_timestamp=is.null&created_at=gte.${since}&select=id,photo_url,fraud_flags,status,event_type&limit=200`,
      { headers: HEADERS }
    );
    if (!res.ok) throw new Error(`Fetch failed: ${res.status}`);

    const events = await res.json();
    console.log(`📦 Found ${events.length} events without timestamp`);

    let updated = 0;

    for (const ev of events) {
      try {
        const photo = await downloadPhoto(ev.photo_url);
        const ocrTime = await extractOcrTimestamp(photo, worker);

        if (!ocrTime) {
          console.log(`  — Event ${ev.id}: OCR found no timestamp`);
          continue;
        }

        if (!isValidDate(ocrTime)) {
          console.warn(`  ⚠️  Event ${ev.id}: invalid/implausible date rejected (${ocrTime})`);
          continue;
        }

        await updateEvent(ev.id, ocrTime, ev.fraud_flags, ev.status);
        updated++;
        console.log(`  ✏️  Event ${ev.id}: → ${ocrTime}`);
      } catch (err) {
        console.warn(`  ⚠️  Event ${ev.id}: ${err.message}`);
      }
    }

    console.log(`✅ OCR Worker: ${updated}/${events.length} events updated`);
  } catch (err) {
    // Не останавливаем pipeline — AI-worker должен запуститься в любом случае
    console.error('❌ OCR Worker failed:', err.message);
  } finally {
    if (worker) await worker.terminate().catch(() => {});
  }
}

// .catch гарантирует что необработанное исключение не даёт exit code 1
main().catch(err => console.error('❌ OCR Worker fatal:', err.message));

File: ai-worker/ocr_worker.py
#!/usr/bin/env python3
"""
OCR Worker — extract photo_timestamp from events using EasyOCR.
Replaces scripts/ocr-worker.js (Tesseract) with EasyOCR for better scene text recognition.

Usage:
  python ai-worker/ocr_worker.py            — recent events (last 4h, max 50)
  python ai-worker/ocr_worker.py --backfill — all events since 2026-05-01, paginated
"""

import sys
import re
import io
import os
import argparse
from datetime import datetime, timezone, timedelta

import numpy as np
from PIL import Image
import requests
import easyocr

SUPABASE_URL = os.environ["SUPABASE_URL"]
SERVICE_KEY  = os.environ["SUPABASE_SERVICE_ROLE_KEY"]

HEADERS = {
    "apikey":        SERVICE_KEY,
    "Authorization": f"Bearer {SERVICE_KEY}",
    "Content-Type":  "application/json",
}

RU_MONTHS = {
    'янв': 1, 'фев': 2, 'мар': 3, 'апр': 4, 'май': 5, 'мая': 5,
    'июн': 6, 'июл': 7, 'авг': 8, 'сен': 9, 'окт': 10, 'ноя': 11, 'дек': 12,
}
EN_MONTHS = {
    'jan': 1, 'feb': 2, 'mar': 3, 'apr': 4, 'may': 5, 'jun': 6,
    'jul': 7, 'aug': 8, 'sep': 9, 'oct': 10, 'nov': 11, 'dec': 12,
}


def prepare_text(raw: str) -> str:
    t = raw
    # OCR шум: ведущая "1" перед временем (" 123:55:30" → " 23:55:30")
    t = re.sub(r'\s1(\d{2}:\d{2}:\d{2})', r' \1', t)
    t = re.sub(r'(\d{2})1(\d{2}):(\d{2})', r'\1:\2:\3', t)
    # Латинские OCR-замены для названий месяцев
    t = re.sub(r'(\d{1,2})\s*mas\s*(\d{4})',  r'\1 мая \2', t, flags=re.IGNORECASE)
    t = re.sub(r'(\d{1,2})\s*mai\s*(\d{4})',  r'\1 мая \2', t, flags=re.IGNORECASE)
    t = re.sub(r'(\d{1,2})\s*map\s*(\d{4})',  r'\1 мар \2', t, flags=re.IGNORECASE)
    t = re.sub(r'(\d{1,2})\s*waa\s*(\d{4})',  r'\1 мая \2', t, flags=re.IGNORECASE)
    t = re.sub(r'(\d{1,2})\s*was\s*(\d{4})',  r'\1 мая \2', t, flags=re.IGNORECASE)
    # Пунктуация после названий месяцев: "июн." "июн," "июн:" "июн;" → "июн "
    t = re.sub(r'([а-яёa-z]{3,4})[!.:;,]\s*', r'\1 ', t, flags=re.IGNORECASE)
    # Убираем "г." "г," "г:" "г;" "г!" после года: "2026 г:" → "2026 "
    t = re.sub(r'(\d{4})\s*[гпrР][.!,:;]\s*', r'\1 ', t)
    # Мусорная цифра/символ сразу после 4-значного года ("20261 " → "2026 ")
    t = re.sub(r'(\d{4})[01]([.\s,;])', r'\1\2', t)
    # Точка сразу после года ("2026." → "2026 ")
    t = re.sub(r'(\d{4})\.', r'\1 ', t)
    # Изолированный "1" или "l" между годом и временем ("2026 1," → "2026 ")
    t = re.sub(r'(\d{4})\s+[1l][,\s]', r'\1 ', t, flags=re.IGNORECASE)
    # Время с точками/запятыми вместо двоеточий: "18.04.11" → "18:04:11"
    # Negative lookahead (?!\d) защищает даты "13.06.2026" от замены
    def _dot_to_colon(m):
        h, mi, s = int(m.group(1)), int(m.group(2)), int(m.group(3))
        if h <= 23 and mi <= 59 and s <= 59:
            return f'{m.group(1)}:{m.group(2)}:{m.group(3)}'
        return m.group(0)
    t = re.sub(r'\b(\d{1,2})[.,](\d{2})[.,](\d{2})(?!\d)', _dot_to_colon, t)
    return t


def try_extract(text: str, created_date: str = None) -> dict:
    t = prepare_text(text)

    # Формат: DD месяц YYYY HH:MM[:SS]
    full = re.search(
        r'(\d{1,2})\s*([а-яёa-z]{3,})\s*(\d{4})[^0-9]+(\d{1,2}):(\d{2})(?::(\d{2}))?',
        t, re.IGNORECASE
    )
    if full:
        key  = full.group(2).lower()[:3]
        mnum = RU_MONTHS.get(key) or EN_MONTHS.get(key)
        if mnum:
            return {
                'found': True, 'day': full.group(1), 'month': mnum,
                'year':  full.group(3), 'h': full.group(4),
                'm':     full.group(5), 's': full.group(6) or '0',
            }

    time_  = re.search(r'(\d{1,2}):(\d{2}):(\d{2})', t)
    timehm = re.search(r'\b(\d{1,2}):(\d{2})\b', t)
    dmy    = re.search(r'(\d{1,2})[./](\d{1,2})[./](\d{2,4})', t)
    ymd    = re.search(r'(\d{4})[-./](\d{2})[-./](\d{2})', t)

    def norm_year(y): return str(2000 + int(y)) if len(y) == 2 else y

    if time_ and dmy:
        return {'found': True, 'day': dmy.group(1), 'month': int(dmy.group(2)),
                'year': norm_year(dmy.group(3)), 'h': time_.group(1), 'm': time_.group(2), 's': time_.group(3)}
    if time_ and ymd:
        return {'found': True, 'day': ymd.group(3), 'month': int(ymd.group(2)),
                'year': ymd.group(1), 'h': time_.group(1), 'm': time_.group(2), 's': time_.group(3)}
    if timehm and dmy:
        return {'found': True, 'day': dmy.group(1), 'month': int(dmy.group(2)),
                'year': norm_year(dmy.group(3)), 'h': timehm.group(1), 'm': timehm.group(2), 's': '0'}
    if timehm and ymd:
        return {'found': True, 'day': ymd.group(3), 'month': int(ymd.group(2)),
                'year': ymd.group(1), 'h': timehm.group(1), 'm': timehm.group(2), 's': '0'}

    # Fallback: год + время (без даты) — берём дату из created_at
    if created_date:
        yt = re.search(r'\b(\d{4})\b[^0-9]*(\d{1,2}):(\d{2})(?::(\d{2}))?', t)
        if yt:
            y, h, m, s = yt.group(1), yt.group(2), yt.group(3), yt.group(4) or '0'
            if y == created_date[:4]:
                return {'found': True, 'day': created_date[8:10], 'month': int(created_date[5:7]),
                        'year': y, 'h': h, 'm': m, 's': s}

    return {'found': False}


def to_iso(r: dict) -> str:
    moscow = datetime(
        int(r['year']), int(r['month']), int(r['day']),
        int(r['h']), int(r['m']), int(r['s']),
        tzinfo=timezone(timedelta(hours=3))
    )
    return moscow.astimezone(timezone.utc).isoformat().replace('+00:00', 'Z')


def is_valid_date(iso_str: str) -> bool:
    try:
        d = datetime.fromisoformat(iso_str.replace('Z', '+00:00'))
    except Exception:
        return False
    now = datetime.now(timezone.utc)
    if d > now + timedelta(hours=6):
        return False
    if d < datetime(2026, 1, 1, tzinfo=timezone.utc):
        return False
    return True


def extract_ocr_timestamp(image_bytes: bytes, reader: easyocr.Reader, created_date: str = None, debug: bool = False):
    try:
        img     = Image.open(io.BytesIO(image_bytes)).convert('RGB')
        img_arr = np.array(img)

        # Pass A: EasyOCR на полном изображении (RGB)
        results = reader.readtext(img_arr)
        text_a  = ' '.join(t for _, t, _ in results)
        if debug:
            print(f'    [DEBUG] Pass A raw: {repr(text_a[:300])}')
            print(f'    [DEBUG] Pass A prepared: {repr(prepare_text(text_a)[:300])}')
        res_a   = try_extract(text_a, created_date)
        if res_a['found']:
            return to_iso(res_a)

        # Pass B: grayscale + повтор
        gray     = np.array(img.convert('L'))
        results2 = reader.readtext(gray)
        text_b   = ' '.join(t for _, t, _ in results2)
        if debug:
            print(f'    [DEBUG] Pass B raw: {repr(text_b[:300])}')
        res_b    = try_extract(text_b, created_date)
        if res_b['found']:
            return to_iso(res_b)

    except Exception as e:
        print(f'    EasyOCR error: {e}')

    return None


def download_photo(photo_url: str) -> bytes:
    url = f"{SUPABASE_URL}/storage/v1/object/{photo_url}"
    r   = requests.get(url, headers=HEADERS, timeout=30)
    if not r.ok:
        raise RuntimeError(f"Storage download failed: {r.status_code}")
    return r.content


def update_event(event_id: int, photo_timestamp: str, current_flags, event_status: str):
    flags = [f for f in (current_flags or []) if f != 'no_photo_time']
    patch = {'photo_timestamp': photo_timestamp, 'fraud_flags': flags}
    if event_status == 'done':
        patch['status'] = 'pending'
    r = requests.patch(
        f"{SUPABASE_URL}/rest/v1/events?id=eq.{event_id}",
        headers=HEADERS, json=patch, timeout=15,
    )
    if not r.ok:
        raise RuntimeError(f"Update failed: {r.status_code}")


def _process_events(events: list, reader: easyocr.Reader, debug: bool = False) -> int:
    updated = 0
    for ev in events:
        try:
            photo       = download_photo(ev['photo_url'])
            moscow_date = (
                datetime.fromisoformat(ev['created_at'].replace('Z', '+00:00'))
                + timedelta(hours=3)
            ).strftime('%Y-%m-%d')

            ocr_time = extract_ocr_timestamp(photo, reader, moscow_date, debug=debug)

            if not ocr_time:
                print(f"  — Event {ev['id']}: no timestamp found")
                continue

            if not is_valid_date(ocr_time):
                print(f"  ⚠️  Event {ev['id']}: invalid date rejected ({ocr_time})")
                continue

            update_event(ev['id'], ocr_time, ev.get('fraud_flags'), ev['status'])
            updated += 1
            print(f"  ✏️  Event {ev['id']}: → {ocr_time}")

        except Exception as e:
            print(f"  ⚠️  Event {ev['id']}: {e}")

    print(f"  Updated {updated}/{len(events)}")
    return updated


def run_regular(reader: easyocr.Reader, debug: bool = False):
    since = (datetime.now(timezone.utc) - timedelta(hours=4)).isoformat().replace('+00:00', 'Z')
    url   = (
        f"{SUPABASE_URL}/rest/v1/events"
        f"?photo_url=not.is.null&photo_timestamp=is.null"
        f"&created_at=gte.{since}"
        f"&select=id,photo_url,fraud_flags,status,event_type,created_at"
        f"&limit=50&order=created_at.desc"
    )
    r = requests.get(url, headers=HEADERS, timeout=30)
    if not r.ok:
        raise RuntimeError(f"Fetch failed: {r.status_code}")
    events = r.json()
    print(f"📦 Found {len(events)} events without timestamp")
    _process_events(events, reader, debug=debug)


def run_debug(reader: easyocr.Reader):
    """Берёт 5 случайных фото без timestamp и печатает что EasyOCR реально видит."""
    url = (
        f"{SUPABASE_URL}/rest/v1/events"
        f"?photo_url=not.is.null&photo_timestamp=is.null"
        f"&select=id,photo_url,fraud_flags,status,event_type,created_at"
        f"&limit=5&order=created_at.desc"
    )
    r = requests.get(url, headers=HEADERS, timeout=30)
    if not r.ok:
        raise RuntimeError(f"Fetch failed: {r.status_code}")
    events = r.json()
    print(f"🔍 DEBUG: проверяю {len(events)} фото")
    _process_events(events, reader, debug=True)


def run_backfill(reader: easyocr.Reader):
    print("📦 Backfill: processing all events without timestamp since 2026-05-01")
    since     = "2026-05-01T00:00:00Z"
    page_size = 1000
    offset    = 0
    total_updated = 0
    total_seen    = 0

    while True:
        url = (
            f"{SUPABASE_URL}/rest/v1/events"
            f"?photo_url=not.is.null&photo_timestamp=is.null"
            f"&created_at=gte.{since}"
            f"&select=id,photo_url,fraud_flags,status,event_type,created_at"
            f"&limit={page_size}&offset={offset}&order=created_at.asc"
        )
        r = requests.get(url, headers=HEADERS, timeout=30)
        if not r.ok:
            raise RuntimeError(f"Fetch page failed: {r.status_code}")

        events = r.json()
        if not events:
            break

        print(f"  Page {offset // page_size + 1}: {len(events)} events")
        updated       = _process_events(events, reader)
        total_updated += updated
        total_seen    += len(events)
        offset        += page_size

        if len(events) < page_size:
            break

    print(f"✅ Backfill done: {total_updated}/{total_seen} events updated")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--backfill', action='store_true',
                        help='Process all events since 2026-05-01')
    parser.add_argument('--debug', action='store_true',
                        help='Show raw EasyOCR output for 5 recent photos (no DB writes)')
    args = parser.parse_args()

    print("🔄 OCR Worker (EasyOCR) started")
    print("  Loading EasyOCR models (ru + en) ...")
    reader = easyocr.Reader(['ru', 'en'], gpu=False)
    print("  Models loaded ✓")

    try:
        if args.debug:
            run_debug(reader)
        elif args.backfill:
            run_backfill(reader)
        else:
            run_regular(reader)
    except Exception as e:
        print(f"❌ OCR Worker failed: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()
